File: ws-node/src/primes.test.ts
// Copyright 2026 TEAM RelicFrog
// SPDX-License-Identifier: Apache-2.0
//
// primes.test.ts — Unit tests for the primes module.
// Uses the built-in node:test runner (no external dependencies).

import assert from "node:assert/strict";
import { describe, it } from "node:test";
import {
  InvalidNthIndexError,
  InvalidRangeError,
  isPrime,
  LimitTooSmallError,
  nthPrime,
  primesInRange,
  sieveOfEratosthenes,
} from "./primes.js";

// ---------------------------------------------------------------------------
// isPrime
// ---------------------------------------------------------------------------

describe("isPrime", () => {
  it("returns false for values below 2", () => {
    assert.equal(isPrime(0n), false);
    assert.equal(isPrime(1n), false);
  });

  it("returns true for 2", () => {
    assert.equal(isPrime(2n), true);
  });

  it("returns true for 3", () => {
    assert.equal(isPrime(3n), true);
  });

  it("returns false for even composites", () => {
    assert.equal(isPrime(4n), false);
    assert.equal(isPrime(100n), false);
    assert.equal(isPrime(1_000_000n), false);
  });

  it("returns true for known primes", () => {
    const known = [2n, 3n, 5n, 7n, 11n, 13n, 17n, 19n, 23n, 29n, 31n, 97n];
    for (const p of known) {
      assert.equal(isPrime(p), true, `${p} should be prime`);
    }
  });

  it("returns false for known composites", () => {
    const known = [4n, 6n, 8n, 9n, 15n, 25n, 49n, 91n, 100n];
    for (const c of known) {
      assert.equal(isPrime(c), false, `${c} should not be prime`);
    }
  });

  it("returns true for a large prime", () => {
    assert.equal(isPrime(999_983n), true);
  });

  it("returns false for a large composite", () => {
    assert.equal(isPrime(999_999n), false); // 3 × 333333
  });
});

// ---------------------------------------------------------------------------
// sieveOfEratosthenes
// ---------------------------------------------------------------------------

describe("sieveOfEratosthenes", () => {
  it("throws LimitTooSmallError for limit < 2", () => {
    assert.throws(() => sieveOfEratosthenes(0n), LimitTooSmallError);
    assert.throws(() => sieveOfEratosthenes(1n), LimitTooSmallError);
  });

  it("returns [2] for limit=2", () => {
    assert.deepEqual(sieveOfEratosthenes(2n), [2n]);
  });

  it("returns primes up to 10", () => {
    assert.deepEqual(sieveOfEratosthenes(10n), [2n, 3n, 5n, 7n]);
  });

  it("returns primes up to 30", () => {
    assert.deepEqual(sieveOfEratosthenes(30n), [2n, 3n, 5n, 7n, 11n, 13n, 17n, 19n, 23n, 29n]);
  });

  it("returns 25 primes for limit=99 (primes below 100)", () => {
    // There are 25 primes <= 99: 2..97
    assert.equal(sieveOfEratosthenes(99n).length, 25);
  });

  it("returns 25 primes for limit=100", () => {
    assert.equal(sieveOfEratosthenes(100n).length, 25);
  });
});

// ---------------------------------------------------------------------------
// primesInRange
// ---------------------------------------------------------------------------

describe("primesInRange", () => {
  it("throws InvalidRangeError when start > end", () => {
    assert.throws(() => primesInRange(10n, 5n), InvalidRangeError);
  });

  it("returns [] when end < 2", () => {
    assert.deepEqual(primesInRange(0n, 1n), []);
  });

  it("returns [7] for range [7, 7]", () => {
    assert.deepEqual(primesInRange(7n, 7n), [7n]);
  });

  it("returns primes in [10, 20]", () => {
    assert.deepEqual(primesInRange(10n, 20n), [11n, 13n, 17n, 19n]);
  });

  it("returns [] for a composite singleton range", () => {
    assert.deepEqual(primesInRange(9n, 9n), []);
  });
});

// ---------------------------------------------------------------------------
// nthPrime
// ---------------------------------------------------------------------------

describe("nthPrime", () => {
  it("throws InvalidNthIndexError for n=0", () => {
    assert.throws(() => nthPrime(0n), InvalidNthIndexError);
  });

  it("returns 2 for n=1", () => {
    assert.equal(nthPrime(1n), 2n);
  });

  it("returns 3 for n=2", () => {
    assert.equal(nthPrime(2n), 3n);
  });

  it("returns 29 for n=10", () => {
    assert.equal(nthPrime(10n), 29n);
  });

  it("returns 541 for n=100", () => {
    assert.equal(nthPrime(100n), 541n);
  });

  it("sequence matches sieve for first 10 primes", () => {
    const sieve = sieveOfEratosthenes(30n);
    for (let i = 0; i < sieve.length; i++) {
      assert.equal(nthPrime(BigInt(i + 1)), sieve[i]);
    }
  });
});

File: ws-lua/src/primes_test.lua
-- Copyright 2026 TEAM RelicFrog
-- SPDX-License-Identifier: Apache-2.0
--
-- src/primes_test.lua — Unit tests for the primes module (LuaJIT).
-- Uses a minimal inline test harness (no external dependencies).

local primes = require("primes")

-- ---------------------------------------------------------------------------
-- Minimal test harness
-- ---------------------------------------------------------------------------

local pass = 0
local fail = 0

local function ok(desc, cond)
	if cond then
		pass = pass + 1
		io.write(string.format("  ok  %d — %s\n", pass + fail, desc))
	else
		fail = fail + 1
		io.write(string.format("  FAIL %d — %s\n", pass + fail, desc))
	end
end

local function raises(desc, fn)
	local ok_flag, _ = pcall(fn)
	ok(desc, not ok_flag)
end

local function eq_list(a, b)
	if #a ~= #b then
		return false
	end
	for i = 1, #a do
		if a[i] ~= b[i] then
			return false
		end
	end
	return true
end

-- ---------------------------------------------------------------------------
-- is_prime
-- ---------------------------------------------------------------------------

ok("is_prime(0) == false", not primes.is_prime(0))
ok("is_prime(1) == false", not primes.is_prime(1))
ok("is_prime(2) == true", primes.is_prime(2))
ok("is_prime(3) == true", primes.is_prime(3))
ok("is_prime(4) == false", not primes.is_prime(4))
ok("is_prime(100) == false", not primes.is_prime(100))
ok("is_prime(97) == true", primes.is_prime(97))
ok("is_prime(999983) == true", primes.is_prime(999983))
ok("is_prime(999999) == false", not primes.is_prime(999999))

for _, p in ipairs({ 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 97 }) do
	ok("is_prime(" .. p .. ")", primes.is_prime(p))
end

for _, c in ipairs({ 4, 6, 8, 9, 15, 25, 49, 91, 100 }) do
	ok("not is_prime(" .. c .. ")", not primes.is_prime(c))
end

-- ---------------------------------------------------------------------------
-- sieve_of_eratosthenes
-- ---------------------------------------------------------------------------

raises("sieve limit=1 raises", function()
	primes.sieve_of_eratosthenes(1)
end)
raises("sieve limit=0 raises", function()
	primes.sieve_of_eratosthenes(0)
end)
raises("sieve limit=-1 raises", function()
	primes.sieve_of_eratosthenes(-1)
end)

ok("sieve(2) == {2}", eq_list(primes.sieve_of_eratosthenes(2), { 2 }))
ok("sieve(10) == {2,3,5,7}", eq_list(primes.sieve_of_eratosthenes(10), { 2, 3, 5, 7 }))
ok("sieve(30) has 10 primes", #primes.sieve_of_eratosthenes(30) == 10)
ok("sieve(99) has 25 primes", #primes.sieve_of_eratosthenes(99) == 25)
ok("sieve(100) has 25 primes", #primes.sieve_of_eratosthenes(100) == 25)

-- ---------------------------------------------------------------------------
-- primes_in_range
-- ---------------------------------------------------------------------------

raises("range start>end raises", function()
	primes.primes_in_range(10, 5)
end)

ok("range(0,1) == {}", eq_list(primes.primes_in_range(0, 1), {}))
ok("range(7,7) == {7}", eq_list(primes.primes_in_range(7, 7), { 7 }))
ok("range(9,9) == {}", eq_list(primes.primes_in_range(9, 9), {}))
ok("range(10,20) correct", eq_list(primes.primes_in_range(10, 20), { 11, 13, 17, 19 }))

-- ---------------------------------------------------------------------------
-- nth_prime
-- ---------------------------------------------------------------------------

raises("nth_prime(0) raises", function()
	primes.nth_prime(0)
end)
raises("nth_prime(-1) raises", function()
	primes.nth_prime(-1)
end)

ok("nth_prime(1) == 2", primes.nth_prime(1) == 2)
ok("nth_prime(2) == 3", primes.nth_prime(2) == 3)
ok("nth_prime(10) == 29", primes.nth_prime(10) == 29)
ok("nth_prime(100) == 541", primes.nth_prime(100) == 541)

local sieve = primes.sieve_of_eratosthenes(30)
for i, expected in ipairs(sieve) do
	ok("nth_prime(" .. i .. ") == " .. expected, primes.nth_prime(i) == expected)
end

-- ---------------------------------------------------------------------------
-- Summary
-- ---------------------------------------------------------------------------

print(string.format("\n%d tests run: %d passed, %d failed", pass + fail, pass, fail))
if fail > 0 then
	os.exit(1)
end
